python/cudf/cudf/utils/_setup_numba.py

```diff
@@ -0,0 +1,157 @@
+    """
+    filtered_archs = [x for x in archs if x[0] <= max_compute_capability]
+    if filtered_archs:
+        return max(filtered_archs, key=lambda x: x[0])
+    else:
+        return None
+
+
+def _get_ptx_file(path, prefix):
+    if "RAPIDS_NO_INITIALIZE" in os.environ:
+        # cc=60 ptx is always built
+        cc = int(os.environ.get("STRINGS_UDF_CC", "60"))
+    else:
+        from numba import cuda
+
+        dev = cuda.get_current_device()
+
+        # Load the highest compute capability file available that is less than
+        # the current device's.
+        cc = int("".join(str(x) for x in dev.compute_capability))
+    files = glob.glob(os.path.join(path, f"{prefix}*.ptx"))
+    if len(files) == 0:
+        raise RuntimeError(f"Missing PTX files for cc={cc}")
+    regular_sms = []
+
+    for f in files:
+        file_name = os.path.basename(f)
+        sm_number = file_name.rstrip(".ptx").lstrip(prefix)
+        if sm_number.endswith("a"):
+            processed_sm_number = int(sm_number.rstrip("a"))
+            if processed_sm_number == cc:
+                return f
+        else:
+            regular_sms.append((int(sm_number), f))
+
+    regular_result = None
+
+    if regular_sms:
+        regular_result = _get_best_ptx_file(regular_sms, cc)
+
+    if regular_result is None:
+        raise RuntimeError(
+            "This cuDF installation is missing the necessary PTX "
+            f"files that are <={cc}."
+        )
+    else:
+        return regular_result[1]
+
+
+def _setup_numba_linker(path):
+    try:
+        # By default, ptxcompiler will not be installed with CUDA 12
+        # packages. This is ok, because in this situation putting
+        # numba in enhanced compatibility mode is not necessary.
+        from ptxcompiler.patch import NO_DRIVER, safe_get_versions
+    except ImportError:
+        return
+    versions = safe_get_versions()
+    if versions != NO_DRIVER:
+        driver_version, runtime_version = versions
+        # Don't check if CEC is necessary in the possible edge
+        # case where a user has a CUDA 12 package and ptxcompiler
+        # in their environment anyways, perhaps installed separately
+        if driver_version < (12, 0):
+            ptx_toolkit_version = _get_cuda_version_from_ptx_file(path)
+            maybe_patch_numba_linker(
+                driver_version, runtime_version, ptx_toolkit_version
+            )
+
+
+def maybe_patch_numba_linker(
+    driver_version, runtime_version, ptx_toolkit_version
+):
+    # Numba thinks cubinlinker is only needed if the driver is older than
+    # the CUDA runtime, but when PTX files are present, it might also need
+    # to patch because those PTX files may be compiled by a CUDA version
+    # that is newer than the driver as well
+    if (driver_version < ptx_toolkit_version) or (
+        driver_version < runtime_version
+    ):
+        config.NUMBA_CUDA_ENABLE_MINOR_VERSION_COMPATIBILITY = 1
```

Comment:
If we are going to pause this PR for numba 0.57 to be available, then I would prefer that we refactor to avoid this kind of "action-at-a-distance" here (rather than as a followup).

OTOH, if that will hold things up too much, let's not.

Reply:
@wence- Great observation -- and thanks @brandon-b-miller for the context. I agree this would be nice to do contextually for the properties that we can control. Only occupancy warnings is sufficient for me.

Reply:
added something in https://github.com/rapidsai/cudf/pull/13337/commits/96b6f01fda82f3dcc074959fc51975d0cdc3c881